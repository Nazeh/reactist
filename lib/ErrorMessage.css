.reactist_error_message {
  padding: 10px 10px 0 10px;
  margin-bottom: 10px;
  background-color: #f8d9d8;
  border: 1px solid #de4c4a;
}
.reactist_error_message p {
  font-size: 0.875rem;
  color: #de4c4a;
  font-weight: 400;
  line-height: 1.7;
  margin: 0 0 10px 0;
}
.reactist_error_message .dismiss_icon {
  height: 20px;
  width: 20px;
  background-repeat: no-repeat;
  background-position: center;
  background-image: url("data:image/svg+xml,%3Csvg width='23px' height='23px' viewBox='0 0 23 23' version='1.1' xmlns='http://www.w3.org/2000/svg' xmlns:xlink='http://www.w3.org/1999/xlink'%3E %3Cdefs%3E %3Crect id='path-1' x='0' y='0' width='23' height='23'%3E%3C/rect%3E %3C/defs%3E %3Cg id='Symbols' stroke='none' stroke-width='1' fill='none' fill-rule='evenodd'%3E %3Cg id='Icon/close'%3E %3Cg id='tick'%3E %3Cmask id='mask-2' fill='white'%3E %3Cuse xlink:href='%23path-1'%3E%3C/use%3E %3C/mask%3E %3Cg id='Bounds'%3E%3C/g%3E %3C/g%3E %3Cpath d='M11.1696275,10.285744 L7.28244992,6.39856644 C6.98515395,6.10127047 6.51277321,6.10376346 6.21987999,6.39665668 C5.92494477,6.6915919 5.9278418,7.16527866 6.22178974,7.45922661 L10.1089673,11.3464041 L6.39856644,15.056805 C6.10127047,15.354101 6.10376346,15.8264817 6.39665668,16.1193749 C6.6915919,16.4143101 7.16527866,16.4114131 7.45922661,16.1174652 L11.1696275,12.4070643 L15.056805,16.2942419 C15.354101,16.5915378 15.8264817,16.5890448 16.1193749,16.2961516 C16.4143101,16.0012164 16.4114131,15.5275296 16.1174652,15.2335817 L12.2302876,11.3464041 L16.2942419,7.28244992 C16.5915378,6.98515395 16.5890448,6.51277321 16.2961516,6.21987999 C16.0012164,5.92494477 15.5275296,5.9278418 15.2335817,6.22178974 L11.1696275,10.285744 Z' id='Close' fill='%23DE4C4A'%3E%3C/path%3E %3C/g%3E %3C/g%3E %3C/svg%3E");
  float: right;
  position: relative;
  top: -30px;
}
.reactist_error_message .dismiss_icon:hover {
  cursor: pointer;
}

