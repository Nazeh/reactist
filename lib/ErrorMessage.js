!function(e,t){"object"==typeof exports&&"object"==typeof module?module.exports=t(require("prop-types"),require("react")):"function"==typeof define&&define.amd?define("reactist",["prop-types","react"],t):"object"==typeof exports?exports.reactist=t(require("prop-types"),require("react")):e.reactist=t(e["prop-types"],e.React)}(window,function(e,t){return function(e){var t={};function r(n){if(t[n])return t[n].exports;var o=t[n]={i:n,l:!1,exports:{}};return e[n].call(o.exports,o,o.exports,r),o.l=!0,o.exports}return r.m=e,r.c=t,r.d=function(e,t,n){r.o(e,t)||Object.defineProperty(e,t,{enumerable:!0,get:n})},r.r=function(e){"undefined"!=typeof Symbol&&Symbol.toStringTag&&Object.defineProperty(e,Symbol.toStringTag,{value:"Module"}),Object.defineProperty(e,"__esModule",{value:!0})},r.t=function(e,t){if(1&t&&(e=r(e)),8&t)return e;if(4&t&&"object"==typeof e&&e&&e.__esModule)return e;var n=Object.create(null);if(r.r(n),Object.defineProperty(n,"default",{enumerable:!0,value:e}),2&t&&"string"!=typeof e)for(var o in e)r.d(n,o,function(t){return e[t]}.bind(null,o));return n},r.n=function(e){var t=e&&e.__esModule?function(){return e.default}:function(){return e};return r.d(t,"a",t),t},r.o=function(e,t){return Object.prototype.hasOwnProperty.call(e,t)},r.p="",r(r.s=16)}({0:function(t,r){t.exports=e},1:function(e,r){e.exports=t},16:function(e,t,r){"use strict";r.r(t);r(17);var n=r(1),o=r.n(n),i=r(0),u=r.n(i);function a(e){return(a="function"==typeof Symbol&&"symbol"==typeof Symbol.iterator?function(e){return typeof e}:function(e){return e&&"function"==typeof Symbol&&e.constructor===Symbol&&e!==Symbol.prototype?"symbol":typeof e})(e)}function c(e,t){for(var r=0;r<t.length;r++){var n=t[r];n.enumerable=n.enumerable||!1,n.configurable=!0,"value"in n&&(n.writable=!0),Object.defineProperty(e,n.key,n)}}function s(e){return(s=Object.setPrototypeOf?Object.getPrototypeOf:function(e){return e.__proto__||Object.getPrototypeOf(e)})(e)}function f(e,t){return(f=Object.setPrototypeOf||function(e,t){return e.__proto__=t,e})(e,t)}function l(e){if(void 0===e)throw new ReferenceError("this hasn't been initialised - super() hasn't been called");return e}function p(e,t,r){return t in e?Object.defineProperty(e,t,{value:r,enumerable:!0,configurable:!0,writable:!0}):e[t]=r,e}var y=function(e){function t(e,r){var n;!function(e,t){if(!(e instanceof t))throw new TypeError("Cannot call a class as a function")}(this,t),p(l(l(n=function(e,t){return!t||"object"!==a(t)&&"function"!=typeof t?l(e):t}(this,s(t).call(this,e,r)))),"_clearTimeout",function(){n.timeout&&clearTimeout(n.timeout)}),p(l(l(n)),"_triggerDelayedHide",function(){n._clearTimeout(),n.timeout=setTimeout(n._hide,n.props.timeout)}),p(l(l(n)),"_hide",function(){n._clearTimeout(),n.setState(function(){return{visible:!1}}),n.props.onHide&&n.props.onHide()});var o=n._isValidMessage(e.message);return o&&n._triggerDelayedHide(),n.state={visible:o},n}return function(e,t){if("function"!=typeof t&&null!==t)throw new TypeError("Super expression must either be null or a function");e.prototype=Object.create(t&&t.prototype,{constructor:{value:e,writable:!0,configurable:!0}}),t&&f(e,t)}(t,o.a.Component),function(e,t,r){t&&c(e.prototype,t),r&&c(e,r)}(t,[{key:"UNSAFE_componentWillReceiveProps",value:function(e){this._isValidMessage(e.message)&&(this.setState(function(){return{visible:!0}}),this._triggerDelayedHide())}},{key:"_isValidMessage",value:function(e){return e&&e.length>0}},{key:"render",value:function(){var e=this.props.message;return!(!this.state.visible||!this._isValidMessage(e))&&o.a.createElement("div",{className:"error_message"},o.a.createElement("p",null,e),o.a.createElement("div",{onClick:this._hide,className:"dismiss_icon"}))}}]),t}();y.displayName="ErrorMessage",y.defaultProps={timeout:2500},y.propTypes={message:u.a.oneOfType([u.a.string,u.a.arrayOf(u.a.node),u.a.node]),timeout:u.a.number,onHide:u.a.func},t.default=y},17:function(e,t,r){}})});
//# sourceMappingURL=ErrorMessage.map