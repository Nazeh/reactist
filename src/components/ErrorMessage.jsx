import './styles/error_message.less'

import React from 'react'
import PropTypes from 'prop-types'

class ErrorMessage extends React.Component {
    constructor(props, context) {
        super(props, context)

        const is_valid_message = this._isValidMessage(props.message)
        if (is_valid_message) {
            this._triggerDelayedHide()
        }
        this.state = { visible: is_valid_message }
    }

    UNSAFE_componentWillReceiveProps(next_props) {
        if (this._isValidMessage(next_props.message)) {
            this.setState(() => ({ visible: true }))
            this._triggerDelayedHide()
        }
    }

    _isValidMessage(message) {
        return message && message.length > 0
    }

    _clearTimeout = () => {
        this.timeout && clearTimeout(this.timeout)
    }

    _triggerDelayedHide = () => {
        this._clearTimeout()
        this.timeout = setTimeout(this._hide, this.props.timeout)
    }

    _hide = () => {
        this._clearTimeout()
        this.setState(() => ({ visible: false }))
        this.props.onHide && this.props.onHide()
    }

    render() {
        const { message } = this.props
        if (!this.state.visible || !this._isValidMessage(message)) {
            return false
        }

        return (
            <div className="error_message">
                <p>{message}</p>
                <div onClick={this._hide} className="dismiss_icon" />
            </div>
        )
    }
}
ErrorMessage.displayName = 'ErrorMessage'
ErrorMessage.defaultProps = {
    timeout: 2500
}
ErrorMessage.propTypes = {
    /** Message to be displayed. This component only renders when message is set. */
    message: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.arrayOf(PropTypes.node),
        PropTypes.node
    ]),
    /** Timeout after the error message disappears (in ms). */
    timeout: PropTypes.number,
    /** Optional callback that is invoked when the error message disappears. */
    onHide: PropTypes.func
}

export default ErrorMessage
