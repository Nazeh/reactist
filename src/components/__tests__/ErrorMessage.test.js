import React from 'react'
import { shallow } from 'enzyme'

import ErrorMessage from '../ErrorMessage'

describe('ErrorMessage', () => {
    it('renders nothing if message is not set', () => {
        const errorMessage = shallow(getErrorMessage())
        expect(errorMessage).toMatchSnapshot()
    })

    it('renders nothing if message is empty', () => {
        const errorMessage = shallow(getErrorMessage({ message: '' }))
        expect(errorMessage).toMatchSnapshot()
    })

    it('renders the given message', () => {
        const errorMessage = shallow(getErrorMessage({ message: 'Oh noes' }))
        expect(errorMessage).toMatchSnapshot()
    })

    it('renders error message after receiving props', () => {
        const errorMessage = shallow(
            getErrorMessage({ message: 'first message' })
        )

        errorMessage.setProps({ message: 'second message' })
        expect(errorMessage).toMatchSnapshot()
    })

    it('renders nothing after receiving an empty new message', () => {
        const errorMessage = shallow(
            getErrorMessage({ message: 'first message' })
        )

        errorMessage.setProps({ message: '' })
        expect(errorMessage).toMatchSnapshot()
    })

    it('hides after reaching timeout', () => {
        jest.useFakeTimers()
        const errorMessage = shallow(getErrorMessage({ message: 'Oh noes' }))

        expect(errorMessage.state().visible).toBe(true)
        jest.runAllTimers()
        expect(errorMessage.state().visible).toBe(false)
    })

    it('calls onHide callback when hiding', () => {
        jest.useFakeTimers()
        const onHideSpy = jest.fn()
        shallow(getErrorMessage({ message: 'Oh noes', onHide: onHideSpy }))

        jest.runAllTimers()
        expect(onHideSpy).toHaveBeenCalled()
    })

    // Helpers ================================================================
    const getErrorMessage = props => <ErrorMessage {...props} />
})
