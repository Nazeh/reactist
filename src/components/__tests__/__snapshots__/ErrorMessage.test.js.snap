// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorMessage renders error message after receiving props 1`] = `
<div
  className="error_message"
>
  <p>
    second message
  </p>
  <div
    className="dismiss_icon"
    onClick={[Function]}
  />
</div>
`;

exports[`ErrorMessage renders nothing after receiving an empty new message 1`] = `""`;

exports[`ErrorMessage renders nothing if message is empty 1`] = `""`;

exports[`ErrorMessage renders nothing if message is not set 1`] = `""`;

exports[`ErrorMessage renders the given message 1`] = `
<div
  className="error_message"
>
  <p>
    Oh noes
  </p>
  <div
    className="dismiss_icon"
    onClick={[Function]}
  />
</div>
`;
